File: ios/Classes/QueueEvent.swift
//
//  QueueEvent.swift
//  FHStepStepCounterUtil
//
//  Created by Long on 27/02/2024.
//

import Foundation
import Flutter

class QueueEvent : NSObject, FlutterStreamHandler{
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    private var eventSink: FlutterEventSink? = nil
    private let eventChannel: FlutterEventChannel
    
    init(eventChannel: FlutterEventChannel) {
        self.eventChannel = eventChannel
    }
    
    func emit(event: Dictionary<String, Any>){
        self.eventSink?(event)
    }
    
    func emitError(exception: Error){
        self.eventSink?(FlutterError(code: "500", message: "error", details: exception))
    }
}

File: android/src/main/kotlin/com/drag/ss/fh_stepcounter/QueueEvent.kt
package com.dragn.pawday_step

import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import java.lang.Exception

class QueueEvent(private val eventChannel: EventChannel) {

    private var eventSink: EventChannel.EventSink? = null
    companion object{
        private var instance: QueueEvent? = null
        fun getInstance(eventChannel: EventChannel): QueueEvent {
            return instance ?: synchronized(this) {
                instance ?: QueueEvent(eventChannel).also { instance = it }
            }
        }
    }



    fun emit(event: HashMap<String, Any>){
        eventSink?.success(event)
    }

    fun emitError(exception: Exception){
        eventSink?.error("500", "error", exception.toString())
    }
    init {
        eventChannel.setStreamHandler(
            object : EventChannel.StreamHandler{
                override fun onListen(arguments: Any?, events: EventSink?) {
                    eventSink = events
                }
                override fun onCancel(arguments: Any?) {
                    eventSink = null
                }

            }
        )
    }
}